### PR TITLE
fix(cairo): fix premature partition deletion

### PR DESCRIPTION
**Pull request by ideoma**
Fix error found in test.

Here O3PartitionPurgeJob started to delete partition folder TableWriter started to write but has not committed yet.

```
2022-06-20T14:06:09.8413192Z 2022-06-20T14:06:09.750311Z I i.q.c.TableWriter switched partition [path='C:\Users\VSSADM~1\AppData\Local\Temp\junit6769260372375505272\dbRoot\x\1970-01-01T00']
2022-06-20T14:06:09.8414889Z 2022-06-20T14:06:09.755786Z I i.q.c.TableWriter switched partition [path='C:\Users\VSSADM~1\AppData\Local\Temp\junit6769260372375505272\dbRoot\x\1970-01-01T01']
2022-06-20T14:06:09.8416724Z 2022-06-20T14:06:09.760754Z I i.q.c.TableWriter switched partition [path='C:\Users\VSSADM~1\AppData\Local\Temp\junit6769260372375505272\dbRoot\x\1970-01-01T02']
2022-06-20T14:06:09.8418402Z 2022-06-20T14:06:09.764904Z I i.q.c.TableWriter switched partition [path='C:\Users\VSSADM~1\AppData\Local\Temp\junit6769260372375505272\dbRoot\x\1970-01-01T03']
...
2022-06-20T14:06:09.8459110Z 2022-06-20T14:06:09.838178Z I i.q.c.O3PartitionPurgeJob processing [table=x]
2022-06-20T14:06:09.9910645Z 2022-06-20T14:06:09.898648Z I i.q.c.TableWriter switched partition [path='C:\Users\VSSADM~1\AppData\Local\Temp\junit6769260372375505272\dbRoot\x\1970-01-01T15']
2022-06-20T14:06:09.9925714Z 2022-06-20T14:06:09.898761Z I i.q.c.O3PartitionPurgeJob purging removed partition directory [ts=1970-01-01T00:00:00.000000Z, nameTxn=-1]
2022-06-20T14:06:09.9927770Z 2022-06-20T14:06:09.899387Z I i.q.c.O3PartitionPurgeJob purged [path=C:\Users\VSSADM~1\AppData\Local\Temp\junit6769260372375505272\dbRoot\x\1970-01-01T00\]
2022-06-20T14:06:09.9951593Z 2022-06-20T14:06:09.899402Z I i.q.c.O3PartitionPurgeJob purging removed partition directory [ts=1970-01-01T01:00:00.000000Z, nameTxn=-1]
2022-06-20T14:06:09.9954379Z 2022-06-20T14:06:09.900282Z I i.q.c.O3PartitionPurgeJob purged [path=C:\Users\VSSADM~1\AppData\Local\Temp\junit6769260372375505272\dbRoot\x\1970-01-01T01\]

```

This change also changes how newly created partitions are named. Partition name is by `<timestamp>.<txn>` and where `txn` is the transaction number where the partition is created.  Previously `.txn` part of the name only used in case the partition is created in Out Of Order commit, overwriting existing partition. Now `.txn` part always added to the partition folder name, even if it is create with in order commit.

The txn part of the partition name is needed to detect that partition was dropped but partition folder left on the disk. Without the version, there is no way to distinguish a partition which was dropped or is currently written by TableWriter and not committed yet.

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 1 / 1 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/O3PartitionPurgeJob.java|1|1|100.00%|

File: core/src/main/java/io/questdb/cairo/O3PartitionPurgeJob.java [hunk 1]
```diff
@@ -136,7 +136,11 @@ private static void processDetachedPartition(
         long lastTxn = txReader.getTxn();
         for (int i = hi - 2, n = lo - 1; i > n; i -= 2) {
             long nameTxn = partitionList.get(i);
-            boolean rangeUnlocked = txnScoreboard.isRangeAvailable(nameTxn, lastTxn);
+
+            // If partition without version (txnVersion is -1 and nameTxn == 0)
+            // We cannot say if it's the partition TableWriter is writing at the moment.
+            // Similarly if the version on disk is .2 (nameTxn == 3) we can only remove it if the lastTxn 4 or higher
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/O3PartitionPurgeJob.java:142]:
```suggestion
            // Similarly, if the version on disk is .2 (nameTxn == 3) we can only remove it if the lastTxn 4 or higher
```

File: core/src/test/java/io/questdb/griffin/O3PartitionPurgeTest.java [hunk 2]
```diff
@@ -460,7 +497,10 @@ public void testPartitionDeletedAsyncAfterDroppedBySql() throws Exception {
                 runPartitionPurgeJobs();
 
                 path.of(engine.getConfiguration().getRoot()).concat("tbl").concat("1970-01-09").concat("x.d").$();
-                Assert.assertFalse(Chars.toString(path), Files.exists(path));
+
+                // We cannot remove the partition, it can be the version which TableWriter writes
+                // Ideally we want a way to say it's safe to delete dropped partition 1970-01-09 but there is no way atm
+                // Assert.assertFalse(Chars.toString(path), Files.exists(path));
```
Comment (bluestreak01) [inline core/src/test/java/io/questdb/griffin/O3PartitionPurgeTest.java:503]: this is a serious flaw, we now leave dropped partitions behind. Lets discuss

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 1 / 1 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/O3PartitionPurgeJob.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 20 / 20 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TxWriter.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TxReader.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3PartitionPurgeJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|4|4|100.00%|

File: core/src/test/java/io/questdb/griffin/O3PartitionPurgeTest.java (hunk 2 above)
Comment (ideoma) [reply core/src/test/java/io/questdb/griffin/O3PartitionPurgeTest.java]: Ok, this is no longer the case, assert is back and dropped partition are deleted in `O3PartitionPurgeJob`

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 20 / 20 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TxWriter.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TxReader.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3PartitionPurgeJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|4|4|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 20 / 20 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TxWriter.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TxReader.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3PartitionPurgeJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|4|4|100.00%|

**Review by bluestreak01 — APPROVED** [review]

**Pull request merged**